The first RemoteLayerTreeDrawingAreaProxy::commitLayerTree after being idle is always considered 'missed'.

mattwoodrow opened the pull request:
#### 7f170efe52d7a390011fb9434c7c7f220ceb03e4
<pre>
The first RemoteLayerTreeDrawingAreaProxy::commitLayerTree after being idle is always considered &apos;missed&apos;.
<a href="https://bugs.webkit.org/show_bug.cgi?id=257005">https://bugs.webkit.org/show_bug.cgi?id=257005</a>
&lt;rdar://problem/109543927&gt;

Reviewed by Simon Fraser.

 Each RemoteLayerTreeDrawingAreaProxy::commitLayerTree IPC messages results in a reply displayDidRefresh message being sent back to the WebProcess (once the display link fires) to trigger the next rendering update.

If the next commitLayerTree isn&apos;t received before the following display link callback, it goes into a &apos;missed commit&apos; state, assuming the rendering update took more than one interval to render. In missed commit state, when the commitLayerTree eventually arrives, the displayDidRefresh reply is sent immediately, rather than waiting for the next display link callback.

This can&apos;t differentiate between a slow rendering update, and an idle WebContent process that decided it didn&apos;t need to do a rendering update at all. When we do resume from idle, we&apos;ll always be in missed commit state, and do a second rendering update (attempt) immediately, for no real benefit.

This patch adds a new &apos;commitLayerTreeNotTriggered&apos; message so the WebContent process can notify the UI process that it isn&apos;t going to do a rendering update, and to go into an idle state.
It also renames DidUpdateMessageState a bit to more accurately match the IPC message/function names that are being waited on.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeNotTriggered):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
(WebKit::RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay):
(WebKit::RemoteLayerTreeDrawingAreaProxy::waitForDidUpdateActivityState):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::displayDidRefresh):

Canonical link: <a href="https://commits.webkit.org/264538@main">https://commits.webkit.org/264538@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43014dca1094c42cf44c8546f50205b9c8b8ec7c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7953 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8134 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9321 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7849 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9944 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7874 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10718 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7817 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8941 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7064 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9430 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6249 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14669 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7412 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7116 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10528 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7609 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6205 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6939 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1891 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11150 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7343 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -456,6 +456,9 @@
         monitorsToNotify.takeAny()->triggerDisplayDidRefresh();
     m_displayRefreshMonitorsToNotify = nullptr;
+
+    if (!m_updateRenderingTimer.isActive())
```
Could `triggerDisplayDidRefresh` just return a bool to say if it's going to do anything, rather than checking the timer?

mattwoodrow replied:
I don't think we easily can, since we're just triggering `displayDidRefresh` on some number of display refresh monitors. One of those might have `RenderingUpdateScheduler` as a client, and that might ask this same drawing area to trigger a rendering update. I don't think we can hardcode the 'did trigger rendering update' into the `DisplayRefreshMonitor` API.

It seems plausible that we could remove some of the abstraction layers here, but that's a separate task (that I am now intrigued with).

mattwoodrow replied:
Following up, the WIP PR #14327 would make this possible.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
```diff
@@ -23,6 +23,7 @@
 messages -> RemoteLayerTreeDrawingAreaProxy : DrawingAreaProxy NotRefCounted {
     void SetPreferredFramesPerSecond(unsigned preferredFramesPerSecond)
     void WillCommitLayerTree(WebKit::TransactionID transactionID)
+    void WontCommitLayerTree()
```
Slightly odd name. What if DrawingArea::displayDidRefresh() IPC sent a reply instead?

mattwoodrow replied:
It certainly can, it's just a message that isn't specific to RLTDA, so I'll have to add an unnecessary async reply param to WC/CoordinatedGraphics.

mattwoodrow replied:
As discussed offline, changed this message name to `commitLayerTreeNotTriggered` instead.

mattwoodrow committed:
The first RemoteLayerTreeDrawingAreaProxy::commitLayerTree after being idle is always considered 'missed'.
https://bugs.webkit.org/show_bug.cgi?id=257005
<rdar://problem/109543927>

Reviewed by Simon Fraser.

 Each RemoteLayerTreeDrawingAreaProxy::commitLayerTree IPC messages results in a reply displayDidRefresh message being sent back to the WebProcess (once the display link fires) to trigger the next rendering update.

If the next commitLayerTree isn't received before the following display link callback, it goes into a 'missed commit' state, assuming the rendering update took more than one interval to render. In missed commit state, when the commitLayerTree eventually arrives, the displayDidRefresh reply is sent immediately, rather than waiting for the next display link callback.

This can't differentiate between a slow rendering update, and an idle WebContent process that decided it didn't need to do a rendering update at all. When we do resume from idle, we'll always be in missed commit state, and do a second rendering update (attempt) immediately, for no real benefit.

This patch adds a new 'commitLayerTreeNotTriggered' message so the WebContent process can notify the UI process that it isn't going to do a rendering update, and to go into an idle state.
It also renames DidUpdateMessageState a bit to more accurately match the IPC message/function names that are being waited on.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeNotTriggered):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
(WebKit::RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay):
(WebKit::RemoteLayerTreeDrawingAreaProxy::waitForDidUpdateActivityState):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::displayDidRefresh):

Canonical link: https://commits.webkit.org/264538@main
